docs/design/bin_pack_fault_domains.md

```diff
@@ -0,0 +1,197 @@
+  The idea is to have a way to group some processes of a FoundationDB cluster together with the same zone ID, this zone ID will be a logical value like "zone-1".
+
+## Proposed Design
+
+The idea would be to use a custom label like `foundationdb.org/distribution-key`.
+As value we will use a prefix for the `BinPack` mode and a value defining in which zone the process should be running.
+We will also set the `FDB_ZONE_ID` environment variable on the Pod level, the default value of this field is currently the hostname.
+In addition to that we change the `PodAntiAffinity` rule.
+The following would be an example of the `affinity` term for a Pod.
+The example assumes that process should be running in zone `0` and we have set `BinPackAll`:
+
+```yaml
+  affinity:
+    # Affinity to schedule Pods with the same distribution key together
+    podAffinity:
+      preferredDuringSchedulingIgnoredDuringExecution:
+      - podAffinityTerm:
+          labelSelector:
+            matchLabels:
+              foundationdb.org/distribution-key: "all-0"
+              foundationdb.org/fdb-cluster-name: example-cluster
+          topologyKey: fault_domain
+        weight: 1
+    # AntiAffinity to not schedule the Pod on the same fault domain where other Pods with a different distribution key are running
+    podAntiAffinity:
+      # The required statement is per default only allowed on the hostname:
+      # https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#an-example-of-a-pod-that-uses-pod-affinity
+      preferredDuringSchedulingIgnoredDuringExecution:
+      - podAffinityTerm:
+          labelSelector:
+            matchExpressions:
+            - key: "foundationdb.org/distribution-key"
+              operator: NotIn
+              values:
+              - "all-0"
+            - key: "foundationdb.org/fdb-cluster-name"
+              operator: In
+              values:
+              - "example-cluster"
+          topologyKey: fault_domain
+        weight: 1
+```
+
+The additional label and the `affinity` term enable the operator to spread Pods across physical fault domains.
+The following code snippet shows a possible implementation of the required structs:
+
+```golang
+type BinPack string
+
+const (
+    // BinPackAll defines that processes independent of their class should be bin packed.
+    BinPackAll BinPack = "All"
+    // BinPackStateful defines that stateful processes should be bin packed and additionally stateless processes.
+    // The prefix will be either "stateful" or "stateless".
+    BinPackStateful BinPack = "Stateful"
+    // BinPackClass only processes of the same class will be bin packed.
+    // As prefix we will use the according process class.
+    BinPackClass BinPack = "Class"
+)
+
+// DistributionConfig
+type DistributionConfig struct {
+    // Enabled defines if the binpacking is enabled or not.
+    // Default: false
+    Enabled *bool
```

Comment:
In the long term, do you think that we want to default this to true, and possibly deprecate the non-bin-packed behavior entirely?